### PR TITLE
introduce multivariate cdf / quantiles

### PR DESCRIPTION
**Pull request by marcocuturi**
The idea comes from the great work of Marc Hallin and colleagues

https://arxiv.org/abs/1412.8434

The implementation essentially relies on entropic maps (Pooladian/Niles Weed) to approximate both CDF and Quantiles Monge maps (forward and backward between input measure and reference/uniform measure).

We can expect performance to degrade with dimension + sensitivity w.r.t. epsilon, and this can be observed in tests (hence fairly loose `atol` values).

### CODE REVIEW

File: src/ott/tools/soft_sort.py [hunk 1]
```diff
@@ -450,6 +451,84 @@ def _quantile(
...
+    target_sampler: Optional[Callable[[jax.random.PRNGKey, Tuple[int, int]],
+                                      jnp.ndarray]] = None,
+    key: Optional[jax.random.PRNGKey] = None,
+    target_num_samples: Optional[int] = None,
+    cost_fn: Optional[jnp.ndarray] = None,
+    epsilon: Optional[float] = None,
+    input_weights: Optional[jnp.ndarray] = None,
+    target_weights: Optional[jnp.ndarray] = None,
+    **kwargs: Any
+) -> [
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:466]: For clarity, I'd use, e.g., `... -> Tuple[Func_t, Func_t]` and define `Func_t = Callable[[jnp.ndarray], jnp.ndarray]`

File: src/ott/tools/soft_sort.py [hunk 2]
```diff
@@ -450,6 +451,84 @@ def _quantile(
...
 
 
+def multiv_cdf_quantile_maps(
+    inputs: jnp.ndarray,
+    targets: Optional[jnp.ndarray] = None,
+    target_sampler: Optional[Callable[[jax.random.PRNGKey, Tuple[int, int]],
+                                      jnp.ndarray]] = None,
+    key: Optional[jax.random.PRNGKey] = None,
+    target_num_samples: Optional[int] = None,
+    cost_fn: Optional[jnp.ndarray] = None,
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:461]: Should be `Optional[costs.CostFn]`

File: src/ott/tools/soft_sort.py [hunk 3]
```diff
@@ -450,6 +451,84 @@ def _quantile(
...
+  measure by default, but can be modified. For consistency, the reference
+  measure should be symmetrically centered around
+  :math:`(\tfrac{1}{2},\cdots,\tfrac{1}{2})` and suppported on :math:`[0,1]^d`.
+
+  The implementation return two entropic map estimators, one for the CDF map,
+  the other for the quantiles map.
+
+  Args:
+    inputs: 2D array of `[n, d]` vectors.
+    targets: 2D array of `[m,d]` vectors. If not passed explicitly, the random
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:482]: \`[m, d]\`

File: src/ott/tools/soft_sort.py [hunk 4]
```diff
@@ -450,6 +451,84 @@ def _quantile(
...
+      sampler specified in `target_sampled` is used to create these samples from
+      reference measure.
+    target_sampler: When ``targets`` is not passed, Callable that takes a
+      ``key`` and shape specified by ``target_num_samples`` to sample a target
+      points.
+    key: `rng` key used in ``target_sampler``
+    target_num_samples: number of points generated in the target distribution.
+    cost_fn: cost function used to compare ``inputs`` and ``targets``. Passed on
+      to instantiate a {class}`~ott.geometry.pointcloud.PointCloud` object. This
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:491]: Use :class:\`~ott.geometry.pointcloud.PointCloud\` (and similarly in other places).

File: src/ott/tools/soft_sort.py [hunk 5]
```diff
@@ -450,6 +451,84 @@ def _quantile(
...
+    input_weights: `[n,]` vector of weights for input measure. Assumed to be
+      uniform by default.
+    target_weights: `[m,]` vector of weights for target measure. Assumed to be
+      uniform by default.
+    kwargs: keyword arguments passed on to the :func:`~ott.solvers.linear.solve`
+      function, which solves the OT problem between ``inputs`` and ``targets``
+      using the Sinkhorn algorithm.
+
+  Returns:
+    Two callables, vector-to-vector mappings. The first callable implements the
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:504]: I'd format this docstring using a list as, e.g.:
```rst
Two callables, vector-to-vector mappings:
- multivariate CDF map, taking values in the range of the reference measure.
- quantile map, going by default from :math:`[0, 1]^d` to the range of the input measure.
```

File: src/ott/tools/soft_sort.py [hunk 6]
```diff
@@ -450,6 +451,84 @@ def _quantile(
   return apply_on_axis(_quantile, inputs, axis, q, weight, **kwargs)
 
 
+def multiv_cdf_quantile_maps(
+    inputs: jnp.ndarray,
+    targets: Optional[jnp.ndarray] = None,
+    target_sampler: Optional[Callable[[jax.random.PRNGKey, Tuple[int, int]],
+                                      jnp.ndarray]] = None,
+    key: Optional[jax.random.PRNGKey] = None,
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:459]: Should be called rng for consistency. Also would type if for now as `jax.random.PRNGKeyArray` (but not crucial given #446 )

File: src/ott/tools/soft_sort.py (hunk 3 above)
Comment (michalk8) [reply src/ott/tools/soft_sort.py:482]: Also consider using :math:\`[m, d]\` (also elsewhere).

File: src/ott/tools/soft_sort.py [hunk 7]
```diff
@@ -450,6 +451,84 @@ def _quantile(
...
+    Two callables, vector-to-vector mappings. The first callable implements the
+    multivariate CDF map, taking values in the range of the reference measure.
+    The second implements the quantile map, going by default from `[0,1]^d` to
+    the range of the input measure.
+
+  Raises:
+    A ValueError in case the input and target have not the same dimension.
+  """
+  n, d = inputs.shape
+  key = jax.random.PRNGKey(0) if key is None else key
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:513]: Use `from ott import utils; rng = utils.default_prng_key(rng)`

File: tests/tools/soft_sort_test.py [hunk 8]
```diff
@@ -86,6 +86,51 @@ def test_sort_batch(self, rng: jax.random.PRNGKeyArray, topk: int):
     np.testing.assert_array_equal(xs.shape, expected_shape)
     np.testing.assert_array_equal(jnp.diff(xs, axis=axis) >= 0.0, True)
 
+  def test_multiv_cdf_quantiles(self):
+    n, d = 512, 3
+    keys = jax.random.split(jax.random.PRNGKey(0), 3)
```
Comment (michalk8) [inline tests/tools/soft_sort_test.py:91]: Please use the `rng` fixture as `def test_multiv_cdf_quantiles(self, rng: jax.random.PRNGKeyArray):`

File: src/ott/tools/soft_sort.py [hunk 9]
```diff
@@ -450,6 +451,84 @@ def _quantile(
   return apply_on_axis(_quantile, inputs, axis, q, weight, **kwargs)
 
 
+def multiv_cdf_quantile_maps(
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:454]: Please add this function to `__all__`

File: src/ott/tools/soft_sort.py [hunk 10]
```diff
@@ -450,6 +451,84 @@ def _quantile(
...
+    input_weights: Optional[jnp.ndarray] = None,
+    target_weights: Optional[jnp.ndarray] = None,
+    **kwargs: Any
+) -> [
+    Callable[[jnp.ndarray], jnp.ndarray], Callable[[jnp.ndarray], jnp.ndarray]
+]:
+  r"""Returns multivariate CDF and quantile maps, given input samples.
+
+  Implements the multivariate generalizations for CDF and quantiles proposed in
+  :cite:`chernozhukov:17`. The reference measure is assumed to be the uniform
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:472]: Please add this to `docs/references.bib`.

File: src/ott/tools/soft_sort.py (hunk 9 above)
Comment (michalk8) [inline src/ott/tools/soft_sort.py:454]: Also please add this to the appropriate place in `docs/tools.rst`.

File: src/ott/tools/soft_sort.py [hunk 11]
```diff
@@ -450,6 +451,84 @@ def _quantile(
...
+
+  Args:
+    inputs: 2D array of `[n, d]` vectors.
+    targets: 2D array of `[m,d]` vectors. If not passed explicitly, the random
+      sampler specified in `target_sampled` is used to create these samples from
+      reference measure.
+    target_sampler: When ``targets`` is not passed, Callable that takes a
+      ``key`` and shape specified by ``target_num_samples`` to sample a target
+      points.
+    key: `rng` key used in ``target_sampler``
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:488]: Missing `.` at the end.

File: src/ott/tools/soft_sort.py [hunk 12]
```diff
@@ -450,6 +451,84 @@ def _quantile(
...
+      using the Sinkhorn algorithm.
+
+  Returns:
+    Two callables, vector-to-vector mappings. The first callable implements the
+    multivariate CDF map, taking values in the range of the reference measure.
+    The second implements the quantile map, going by default from `[0,1]^d` to
+    the range of the input measure.
+
+  Raises:
+    A ValueError in case the input and target have not the same dimension.
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:510]: This is false, raises :class:\`AssertionError\`

File: src/ott/tools/soft_sort.py [hunk 13]
```diff
@@ -450,6 +451,84 @@ def _quantile(
...
+  n, d = inputs.shape
+  key = jax.random.PRNGKey(0) if key is None else key
+  if targets is None:
+    target_num_samples = n if target_num_samples is None else target_num_samples
+    if target_sampler is None:
+      target_sampler = jax.random.uniform
+    targets = target_sampler(key, (target_num_samples, d))
+  else:
+    d_t = targets.shape[-1]
+    assert d_t == d, f"Incompatible dimensions for inputs/targets, {d_t}≠{d}."
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:521]: Would avoid Unicode symbols and just use `!=`

File: src/ott/tools/soft_sort.py [hunk 14]
```diff
@@ -450,6 +451,84 @@ def _quantile(
...
+    multivariate CDF map, taking values in the range of the reference measure.
+    The second implements the quantile map, going by default from `[0,1]^d` to
+    the range of the input measure.
+
+  Raises:
+    A ValueError in case the input and target have not the same dimension.
+  """
+  n, d = inputs.shape
+  key = jax.random.PRNGKey(0) if key is None else key
+  if targets is None:
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:514]: Per our discussion, I'd prefer one of the following:
- not being able to pass the `target_sampler` and always use `jax.random.uniform` when `targets` are not specified
- remove the `targets` and always require `target_sampler` to be present with the given signature

File: src/ott/tools/soft_sort.py (hunk 12 above)
Comment (michalk8) [reply src/ott/tools/soft_sort.py:510]: Also I think this can be removed from here (would keep the assertion later though).

File: tests/tools/soft_sort_test.py [hunk 15]
```diff
@@ -86,6 +86,51 @@ def test_sort_batch(self, rng: jax.random.PRNGKeyArray, topk: int):
...
+
+    # Set central point in sampled input measure
+    z = jax.random.uniform(keys[0], (1, d))
+
+    # Sample inputs symmetrically centered on z
+    inputs = .34 * jax.random.normal(keys[0], (n, d)) + z
```
Comment (michalk8) [inline tests/tools/soft_sort_test.py:97]: I wanna switch from removing the `0` in `0.34`, could you please add them here and elsewhere? I think it's more legible/less prone to mistakes.

File: src/ott/tools/soft_sort.py [hunk 16]
```diff
@@ -450,6 +451,84 @@ def _quantile(
...
+    targets = target_sampler(key, (target_num_samples, d))
+  else:
+    d_t = targets.shape[-1]
+    assert d_t == d, f"Incompatible dimensions for inputs/targets, {d_t}≠{d}."
+  geom = pointcloud.PointCloud(
+      inputs, targets, cost_fn=cost_fn, epsilon=epsilon
+  )
+  out = linear.solve(geom, a=input_weights, b=target_weights, **kwargs)
+  potentials = out.to_dual_potentials()
+  return potentials.transport, functools.partial(
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:527]: Maybe we can wrap this in some `NamedTuple` object that has `cdf` and `quantile` properties.

File: tests/tools/soft_sort_test.py [hunk 17]
```diff
@@ -86,6 +86,51 @@ def test_sort_batch(self, rng: jax.random.PRNGKeyArray, topk: int):
...
+    np.testing.assert_allclose(z, qua(q), atol=atol)
+
+    # Check consistency with passing sampler / key
+    cdf_2, qua_2 = soft_sort.multiv_cdf_quantile_maps(
+        inputs,
+        target_sampler=jax.random.uniform,
+        key=keys[1],
+        target_num_samples=m
+    )
+    np.testing.assert_allclose(cdf_2(z), cdf(z))
```
Comment (michalk8) [inline tests/tools/soft_sort_test.py:124]: Please add tolerances here.

File: tests/tools/soft_sort_test.py [hunk 18]
```diff
@@ -86,6 +86,51 @@ def test_sort_batch(self, rng: jax.random.PRNGKeyArray, topk: int):
...
+    np.testing.assert_allclose(qua_2(q), qua(q))
```
Comment (michalk8) [inline tests/tools/soft_sort_test.py:125]: Please add tolerances here.

**Review by michalk8 — CHANGES_REQUESTED** [review]

File: src/ott/tools/soft_sort.py (hunk 14 above)
Comment (marcocuturi) [reply src/ott/tools/soft_sort.py:514]: agree, going with the second, because closer, philosophically, to the idea that what we need is a reference measure (i.e. sampler), not an arbitrary point cloud.

File: src/ott/tools/soft_sort.py (hunk 16 above)
Comment (marcocuturi) [reply src/ott/tools/soft_sort.py:527]: maybe overkill? this is likely to only be two Callables, and won't grow beyond that. I can imagine users might call this often as `cdf_map, _ = ...` or the other way round, depending on their need.

File: tests/tools/soft_sort_test.py (hunk 17 above)
Comment (marcocuturi) [reply tests/tools/soft_sort_test.py:124]: no longer relevant, but here default tolerance was good, because the output should be the same (up to numerical precision of rerunning twice the same)

File: tests/tools/soft_sort_test.py (hunk 18 above)
Comment (marcocuturi) [reply tests/tools/soft_sort_test.py:125]: same as above!

File: src/ott/tools/soft_sort.py (hunk 5 above)
Comment (marcocuturi) [reply src/ott/tools/soft_sort.py:504]: thanks!

File: src/ott/tools/soft_sort.py [hunk 19]
```diff
@@ -453,19 +455,16 @@ def _quantile(
...
+    num_target_samples: Optional[int] = None,
+    cost_fn: Optional[costs.CostFn] = None,
     epsilon: Optional[float] = None,
     input_weights: Optional[jnp.ndarray] = None,
     target_weights: Optional[jnp.ndarray] = None,
     **kwargs: Any
-) -> [
-    Callable[[jnp.ndarray], jnp.ndarray], Callable[[jnp.ndarray], jnp.ndarray]
-]:
+) -> [Func_t, Func_t]:
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:467]: Should be `Tuple[Func_t, Func_t]`

File: src/ott/tools/soft_sort.py [hunk 20]
```diff
@@ -478,47 +477,46 @@ def multiv_cdf_quantile_maps(
...
+    inputs: 2D array of :math:`[n, d]` vectors.
+    target_sampler: Callable that takes a ``key`` and ``[m,d]`` shape Tuple.
+      ``m`` is passed on as ``target_num_samples``, dimension ``d`` is inferred
+      directly from the shape passed in ``inputs``. This is assumed by default
+      to be :func:`jax.random.uniform`, and could be any other random sampler
+      properly wrapped to have the signature above.
+    key: rng key used by ``target_sampler``
+    num_target_samples: number ``m`` of points generated in the target
+      distribution.
+    cost_fn: :class:`~ott.geometry.costs.CostFn` object, used to compare
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:489]: I wouldn't use the :class:\`~ott.geometry.costs.CostFn\`, it's already in the type annotation.

File: src/ott/tools/soft_sort.py [hunk 21]
```diff
@@ -478,47 +477,46 @@ def multiv_cdf_quantile_maps(
...
-    cost_fn: cost function used to compare ``inputs`` and ``targets``. Passed on
-      to instantiate a {class}`~ott.geometry.pointcloud.PointCloud` object. This
-      will default to squared-Euclidean distance.
+    inputs: 2D array of :math:`[n, d]` vectors.
+    target_sampler: Callable that takes a ``key`` and ``[m,d]`` shape Tuple.
+      ``m`` is passed on as ``target_num_samples``, dimension ``d`` is inferred
+      directly from the shape passed in ``inputs``. This is assumed by default
+      to be :func:`jax.random.uniform`, and could be any other random sampler
+      properly wrapped to have the signature above.
+    key: rng key used by ``target_sampler``
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:486]: Missing `.` at the end.

File: src/ott/tools/soft_sort.py [hunk 22]
```diff
@@ -478,47 +477,46 @@ def multiv_cdf_quantile_maps(
...
-    target_sampler: When ``targets`` is not passed, Callable that takes a
-      ``key`` and shape specified by ``target_num_samples`` to sample a target
-      points.
-    key: `rng` key used in ``target_sampler``
-    target_num_samples: number of points generated in the target distribution.
-    cost_fn: cost function used to compare ``inputs`` and ``targets``. Passed on
-      to instantiate a {class}`~ott.geometry.pointcloud.PointCloud` object. This
-      will default to squared-Euclidean distance.
+    inputs: 2D array of :math:`[n, d]` vectors.
+    target_sampler: Callable that takes a ``key`` and ``[m,d]`` shape Tuple.
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:481]: Would remove `Tuple`.

File: src/ott/tools/soft_sort.py [hunk 23]
```diff
@@ -478,47 +477,46 @@ def multiv_cdf_quantile_maps(
...
+      ``m`` is passed on as ``target_num_samples``, dimension ``d`` is inferred
+      directly from the shape passed in ``inputs``. This is assumed by default
+      to be :func:`jax.random.uniform`, and could be any other random sampler
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:484]: Would use `~`.

File: src/ott/tools/soft_sort.py [hunk 24]
```diff
@@ -478,47 +477,46 @@ def multiv_cdf_quantile_maps(
...
       function, which solves the OT problem between ``inputs`` and ``targets``
       using the Sinkhorn algorithm.
 
   Returns:
-    Two callables, vector-to-vector mappings. The first callable implements the
-    multivariate CDF map, taking values in the range of the reference measure.
-    The second implements the quantile map, going by default from `[0,1]^d` to
-    the range of the input measure.
+    Two callables, vector-to-vector mappings:
+      - multivariate CDF map, taking values in the range of the reference
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:505]: I don't think this needs to indented, let's see how it renders.

File: src/ott/tools/soft_sort.py (hunk 22 above)
Comment (michalk8) [reply src/ott/tools/soft_sort.py:481]: And would use :math:\`[m, d]\`

File: src/ott/tools/soft_sort.py [hunk 25]
```diff
@@ -478,47 +477,46 @@ def multiv_cdf_quantile_maps(
...
+      directly from the shape passed in ``inputs``. This is assumed by default
+      to be :func:`jax.random.uniform`, and could be any other random sampler
+      properly wrapped to have the signature above.
+    key: rng key used by ``target_sampler``
+    num_target_samples: number ``m`` of points generated in the target
+      distribution.
+    cost_fn: :class:`~ott.geometry.costs.CostFn` object, used to compare
+      ``inputs`` and ``targets``. Passed on to instantiate a
+      :class:`~ott.geometry.pointcloud.PointCloud` object. This
+      defaults to the squared-Euclidean distance.
```
Comment (michalk8) [inline src/ott/tools/soft_sort.py:492]: Would rephrase as:
```rst
If :obj:`None`, :class:`~ott.geometry.costs.SqEuclidean` is used.
```

**Review by michalk8 — CHANGES_REQUESTED** [review]

File: src/ott/tools/soft_sort.py (hunk 22 above)
Comment (marcocuturi) [reply src/ott/tools/soft_sort.py:481]: went for double quote, per our discussion

File: src/ott/tools/soft_sort.py [hunk 26]
```diff
@@ -499,31 +500,33 @@ def multivariate_cdf_quantile_maps(
       to be uniform by default.
     kwargs: keyword arguments passed on to the :func:`~ott.solvers.linear.solve`
       function, which solves the OT problem between ``inputs`` and ``targets``
-      using the Sinkhorn algorithm.
+      using the :class:`~ott.solvers.linear.sinkhorn.Sinkhorn` algorithm.
```
Comment (marcocuturi) [inline src/ott/tools/soft_sort.py:503]: here i wasn't sure about the reference, because we use `solve`... that being said, probably not a good idea to use `LR` on this, because would crash :) so maybe an instance where we should force kwargs to only refer to sinkhorn...

File: src/ott/tools/soft_sort.py [hunk 27]
```diff
@@ -479,12 +480,12 @@ def multivariate_cdf_quantile_maps(
 
   Args:
     inputs: 2D array of ``[n, d]`` vectors.
-    target_sampler: Callable that takes a ``key`` and ``[m,d]`` shape.
+    target_sampler: Callable that takes a ``rng`` and ``[m, d]`` shape.
```
Comment (marcocuturi) [inline src/ott/tools/soft_sort.py:483]: sorry about this one!!!

**Comment by marcocuturi** [comment]
thanks a lot for the last fixes Michal!
